mount: Make snapshots dir structure customizable

aawsome opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Makes the directory structure for displaying/grouping/sorting snapshots customizable for users.
Allows to give arbitrary directory templates containing `%i` (will be replaced by short ID), `%I` (replaced by long ID), `%u` (replaced by username), `%h` (replaced by hostname), `%t` (replaced by tag) and `%T` replaced by the time format.

Also allows to give arbitrary paths within the the directory template and the time format template to make user-defined subdirs possible.

The default behavior can be obtained by specifying the following directory templates:
```
"ids/%i"
"snapshots/%T"
"hosts/%h/%T"
"tags/%t/%T"
```
(which is the default if not explicitly specifying `--path-template`)

Have a look at the unit tests to see what is possible with the new templates.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

The idea comes from #2907.

closes #2907

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

Note to reviewers:  `internal/fuse/snapshots_dir.go` is completely reimplemented; so better not look at the diffs when checking this file.

aawsome committed:
Generalize fuse snapshot dirs implemetation

+ allow "/" in tags and snapshot template

aawsome committed:
Make snapshots dirs in mount command customizable

aawsome commented:
rebased after #3035 has been merged

greatroar commented on internal/fuse/snapshots_dir.go:
```diff
@@ -15,186 +17,183 @@ import (
-	root    *Root
-	names   map[string]*restic.Snapshot
-	snCount int
+// NewSnapshotsDir returns a new directory containing snapshots.
```
Stale comment.

aawsome replied:
is fixed.

greatroar commented:
The reduction in code size is nice, but I don't think a PR that introduces a new formatting language is ready for review without unit tests.

greatroar commented on internal/fuse/snapshots_dir.go:
```diff
@@ -15,186 +17,183 @@ import (
+// If pattern does not exist in a string, the string remains unchanged.
+// If pattern exists, for each repl string a new path with replaced patterns will be added
+// Note that if the pattern exists in a string and repl is empty, the string will be removed.
+func replaceMultiple(paths []string, pattern string, repl []string) (replacedPaths []string) {
```
I find the logic that uses this function hard to follow. Why not have a function that formats a single filename, by just looping over the content?

```go
func formatPath(template string, sn *restic.Snapshot) string {
	var (
		inVerb = false
		out    strings.Builder
	)
	for _, c := range template {
		if !inVerb {
			if c == '%' {
				inVerb = true
			} else {
				out.WriteRune(c)
			}
			continue
		}

		switch c {
		// handle the various formatting verbs
		}
		inVerb = false
	}

	return out.String()
}
```

aawsome replied:
Thanks for the suggestion! It is however a bit more complicated: There can be multiple tags resulting in more than one path and the "latest" link should be handled. I hence extended your code example a bit...
Hope it is now much clearer!

aawsome committed:
Add possibility to set snapshot ID (used in test)

aawsome commented:
@greatroar Thanks for your feedback and suggestions! And sorry for the long delay to improve this. I changed the things as you proposed and added unit tests.

MichaelEischer reviewed (changes requested):
I like the simpler and much more flexible structure provided by this PR. I'm currently a bit worried whether the current implementation could cause some performance problems when navigating repositories with lots of snapshots.

MichaelEischer committed:
fuse: cleanup test

MichaelEischer committed:
fuse: remove unused MetaDir

MichaelEischer committed:
fuse: Redesign snapshot dirstruct

Cleanly separate the directory presentation and the snapshot directory
structure. SnapshotsDir now translates the dirStruct into a format
usable by the fuse library and contains only minimal special case rules.
All decisions have moved into SnapshotsDirStructure which now creates a
fully preassembled tree data structure.

MichaelEischer commented:
I've extended SnapshotsDirStructure to assemble a full fledged tree of meta directories. This simplifies directory accesses to a simple lookup and also moves the logic to handle a `latest` directory out of the `SnapshotsDir`. In total the changes address all my review comments.

MichaelEischer committed:
Mention --snapshot-template and --time-template in changelog

MichaelEischer reviewed (approved):
LGTM. Thanks for the idea to clean up the snapshot dir structure!
